src/kafka-handler.ts

```diff
@@ -1,126 +1,247 @@
+      console.log("Creating Kafka producer...");
+      let client = new kafka.Client(this.config.broker.host, "iotagent-json-producer-" +  Math.floor(Math.random() * 10000));
+      console.log("... Kafka client for producer is ready.");
+      this.producer = new kafka.HighLevelProducer(client, { requireAcks: 1 });
+
+      // Kafka producer events registration
+      this.producer.on("ready", () => {
+        console.log("... Kafka producer creation finished.");
+        this.isProducerReady = true;
+      });
+      this.producer.on("error", (e) => {
+        if (isScheduled == true) {
+          console.log("An operation was already scheduled. No need to do it again.");
+          return;
+        }
+        this.producer.close();
+        isScheduled = true;
+        console.error("Error: ", e);
+        console.log("Will try again to create Kafka producer in a few seconds.");
           setTimeout(() => {
             console.log("Trying again.");
-            this.initKafkaConfiguration(config, callback, "consumer");
-          }, 10000)
-        });
+            this.initDataProducer();
+          }, 10000);
+      });
+      console.log("... Kafka producer created and callbacks registered.");
+    }
+  }
 
-        console.log("... Kafka consumer created and callbacks registered.");
-        break;
+  /**
+   * Generates a JWT for usage on internal requests
+   * @param  tenant Tenancy context to be used
+   * @return        JWT token string
+   */
+  private generateJWT(tenant: string) {
+    const payload = {
+      'service': tenant,
+      'username': 'iotagent'
     }
+    return (new Buffer('dummy jwt schema').toString('base64')) + '.'
+           + (new Buffer(JSON.stringify(payload)).toString('base64')) + '.'
+           + (new Buffer('dummy signature').toString('base64'));
   }
 
   /**
-   * 
-   * @param config The configuration to be used.
-   * @param callback A callback that will process received device manager notifications
+   * Retrieves managed topic id based on given tenant and subject
+   * @param  tenant   Tenant for which a topic is to be retrieved
+   * @param  subject  Subject of the topic to be retrieved
+   * @param  callback Who to call once a topic is obtained
+   * @return void
    */
-  constructor(config: config.ConfigOptions, callback: (data: DeviceManagerEvent) => void) {
-    // Block any communication before producer is properly created.
-    this.isProducerReady = false;
-    this.initKafkaConfiguration(config, callback, "producer");
+  private getPublishTopic(tenant: string, subject: string,
+                          callback: (err?: any, topic?:string) => void,
+                          global?:boolean): void {
+    if (tenant in this.topicMap) {
+      if (subject in this.topicMap[tenant]) {
+        callback(undefined, this.topicMap[tenant][subject]);
+        return;
+      }
+    }
+
+    axios({
+      'url': this.contextBroker + '/topic/' + subject + (global ? "?global=true" : ""),
```

Comment:
I assume that these 'global' topics will be the same for all tenants, right?

Reply:
Precisely.